src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +73,209 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
 {
-  // Mean and variance over the entire training set will be used to compute
-  // the forward pass when deterministic is set to true.
-  if (deterministic)
-  {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
-  }
-  else
-  {
-    mean = arma::mean(input, 1);
-    variance = arma::var(input, 1, 1);
+  Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+      by feature maps.");
+
+  const size_t batchSize = input.n_cols;
 
-    // Normalize the input.
-    output = input.each_col() - mean;
-    inputMean = output;
-    output.each_col() /= arma::sqrt(variance + eps);
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
 
-    // Use Welford method to compute the sample variance and mean.
-    for (size_t i = 0; i < input.n_cols; ++i)
+  // We will calculate minibatch norm on each channel / feature map.
+  if (!deterministic)
+  {
+    // Input size refers to inputWidth * inputHeight.
+    // If inputSize equals one, input corresponds to output from linear layer.
+    const size_t inputSize = input.n_rows / size;
+
+    // Linear output.
+    if (inputSize == 1)
     {
-      count += 1;
+      mean = arma::mean(input, 1);
+      variance = arma::var(input, 1, 1);
+
+      // Normalize the input.
+      output = input.each_col() - mean;
+      inputMean = output;
+      output.each_col() /= arma::sqrt(variance + eps);
+
+      // Use Welford method to compute the sample variance and mean.
+      for (size_t i = 0; i < batchSize; i++)
+      {
+        count += 1;
+
+        OutputDataType diff = input.col(i) - runningMean;
+        if (average)
+        {
+          runningMean = runningMean + diff / count;
+          runningVariance += diff % (input.col(i) - runningMean);
```

Comment:
This implementation would produce slightly different variance values from the one below. for cnn 
Here when we use `runningVariance += diff % (input.col(i) - runningMean);` we are using the updated value of `runningMean` whereas for cnn we have `runningVariance += (diff.t() % diff.t());` . Can you check which one of these is correct ?

Reply:
Also the `runningMean` would be different because we do add iteratively here (keep incrementing count 1 by 1) whereas for cnn we are adding the entire batchsize to count at once.

Reply:
Hey @saksham189, I made the implementation match from PyTorch in mini batchnorm (They don't use Welford to calculate cumulative average / variance). This is much more simplified and I have also added tests for average method and they match to 1e-3 precision.

Reply:
sounds good.